File: output/GPT4o-Mini/xCodeEval_remove_4000/Java/Go/9dd17377edc379115587e50dcf6a8a09.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

type Position struct {
	row  int
	id   int
	last *Position
	next *Position
	link *Position
}

func (p *Position) Compare(o *Position) int {
	if p.row != o.row {
		return p.row - o.row
	}
	return p.id - o.id
}

func (p *Position) Value(r int) int64 {
	if p.link == nil {
		return 0
	}
	return int64(p.row-p.last.row) * int64(r-p.link.row)
}

type TaskE struct{}

func (t *TaskE) Solve(testNumber int, in *InputReader, out *OutputWriter) {
	r := in.ReadInt()
	c := in.ReadInt()
	n := in.ReadInt()
	k := in.ReadInt()
	x := make([]int, n)
	y := make([]int, n)
	id5(in, x, y)
	id4(x, y)
	orderBy(y, x)
	headPosition := make([]*Position, k)
	for i := 0; i < k; i++ {
		headPosition[i] = &Position{-1, -1, nil, nil, nil}
		if i != 0 {
			headPosition[i].last = headPosition[i-1]
			headPosition[i-1].next = headPosition[i]
		}
	}
	tailPosition := make([]*Position, k)
	for i := 0; i < k; i++ {
		tailPosition[i] = &Position{r, -2, nil, nil, nil}
		if i != 0 {
			tailPosition[i].last = tailPosition[i-1]
			tailPosition[i-1].next = tailPosition[i]
		}
	}
	positions := make([]*Position, n)
	for i := 0; i < n; i++ {
		positions[i] = &Position{x[i], i, nil, nil, nil}
	}
	var answer int64
	for i := 0; i < c; i++ {
		at := 0
		for at < n && y[at] < i {
			at++
		}
		for j := 0; j < k-1; j++ {
			headPosition[j+1].link = tailPosition[j]
		}
		set := make(map[*Position]struct{})
		set[headPosition[k-1]] = struct{}{}
		set[tailPosition[0]] = struct{}{}
		headPosition[k-1].next = tailPosition[0]
		tailPosition[0].last = headPosition[k-1]
		var delta int64
		for j := i; j < c; j++ {
			for at < n && y[at] == j {
				var where *Position
				for pos := range set {
					if where == nil || pos.Compare(positions[at]) < 0 {
						where = pos
					}
				}
				current := where
				for l := 1; l < k; l++ {
					delta -= current.Value(r)
					current = current.last
				}
				next := where.next
				delta -= next.Value(r)
				current = positions[at]
				where.next = current
				current.last = where
				next.last = current
				current.next = next
				if k == 1 {
					current.link = current
				} else {
					current.link = where.link
				}
				set[current] = struct{}{}
				delta += next.Value(r)
				delta += current.Value(r)
				current = where
				for l := 1; l < k; l++ {
					current.link = current.link.last
					delta += current.Value(r)
					current = current.last
				}
				at++
			}
			answer += delta
		}
	}
	out.PrintLine(answer)
}

func id5(in *InputReader, arrays ...[]int) {
	for i := 0; i < len(arrays[0]); i++ {
		for j := range arrays {
			arrays[j][i] = in.ReadInt()
		}
	}
}

func id4(arrays ...[]int) {
	for _, array := range arrays {
		for i := range array {
			array[i]--
		}
	}
}

func orderBy(base []int, arrays ...[]int) {
	order := createOrder(len(base))
	sort.Slice(order, func(i, j int) bool {
		return base[order[i]] < base[order[j]]
	})
	for _, array := range arrays {
		temp := make([]int, len(array))
		for i, idx := range order {
			temp[i] = array[idx]
		}
		copy(array, temp)
	}
}

func createOrder(size int) []int {
	order := make([]int, size)
	for i := range order {
		order[i] = i
	}
	return order
}

type InputReader struct {
	scanner *bufio.Scanner
}

func NewInputReader() *InputReader {
	return &InputReader{scanner: bufio.NewScanner(os.Stdin)}
}

func (in *InputReader) ReadInt() int {
	in.scanner.Scan()
	var x int
	fmt.Sscanf(in.scanner.Text(), "%d", &x)
	return x
}

type OutputWriter struct {
	writer *bufio.Writer
}

func NewOutputWriter() *OutputWriter {
	return &OutputWriter{writer: bufio.NewWriter(os.Stdout)}
}

func (out *OutputWriter) PrintLine(i int64) {
	fmt.Fprintln(out.writer, i)
}

func (out *OutputWriter) Close() {
	out.writer.Flush()
}

func main() {
	in := NewInputReader()
	out := NewOutputWriter()
	solver := TaskE{}
	solver.Solve(1, in, out)
	out.Close()
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/Java/C++/9dd17377edc379115587e50dcf6a8a09.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>
#include <stdexcept>
#include <iterator>

using namespace std;

class Position {
public:
    int row;
    int id;
    Position* last;
    Position* next;
    Position* link;

    Position(int row, int id) : row(row), id(id), last(nullptr), next(nullptr), link(nullptr) {}

    bool operator<(const Position& o) const {
        if (row != o.row) {
            return row < o.row;
        }
        return id < o.id;
    }

    long value(int r) {
        if (link == nullptr) {
            return 0;
        }
        return (row - last->row) * (r - link->row);
    }
};

class TaskE {
public:
    void solve(int testNumber, istream& in, ostream& out) {
        int r, c, n, k;
        in >> r >> c >> n >> k;
        vector<int> x(n), y(n);
        for (int i = 0; i < n; i++) {
            in >> x[i];
        }
        for (int i = 0; i < n; i++) {
            in >> y[i];
        }

        for (int i = 0; i < n; i++) {
            x[i]--;
            y[i]--;
        }

        vector<int> order(n);
        iota(order.begin(), order.end(), 0);
        sort(order.begin(), order.end(), [&](int a, int b) {
            return y[a] < y[b];
        });

        vector<Position*> headPosition(k);
        for (int i = 0; i < k; i++) {
            headPosition[i] = new Position(-1, -1);
            if (i != 0) {
                headPosition[i]->last = headPosition[i - 1];
                headPosition[i - 1]->next = headPosition[i];
            }
        }

        vector<Position*> tailPosition(k);
        for (int i = 0; i < k; i++) {
            tailPosition[i] = new Position(r, -2);
            if (i != 0) {
                tailPosition[i]->last = tailPosition[i - 1];
                tailPosition[i - 1]->next = tailPosition[i];
            }
        }

        vector<Position*> positions(n);
        for (int i = 0; i < n; i++) {
            positions[i] = new Position(x[i], i);
        }

        long answer = 0;
        for (int i = 0; i < c; i++) {
            int at = 0;
            while (at < n && y[at] < i) {
                at++;
            }
            for (int j = 0; j < k - 1; j++) {
                headPosition[j + 1]->link = tailPosition[j];
            }
            set<Position*> setPositions;
            setPositions.insert(headPosition[k - 1]);
            setPositions.insert(tailPosition[0]);
            headPosition[k - 1]->next = tailPosition[0];
            tailPosition[0]->last = headPosition[k - 1];
            long delta = 0;

            for (int j = i; j < c; j++) {
                while (at < n && y[at] == j) {
                    Position* where = *setPositions.lower_bound(positions[at]);
                    Position* current = where;
                    for (int l = 1; l < k; l++) {
                        delta -= current->value(r);
                        current = current->last;
                    }
                    Position* next = where->next;
                    delta -= next->value(r);
                    current = positions[at];
                    where->next = current;
                    current->last = where;
                    next->last = current;
                    current->next = next;
                    current->link = (k == 1) ? current : where->link;
                    setPositions.insert(current);
                    delta += next->value(r);
                    delta += current->value(r);
                    current = where;
                    for (int l = 1; l < k; l++) {
                        current->link = current->link->last;
                        delta += current->value(r);
                        current = current->last;
                    }
                    at++;
                }
                answer += delta;
            }
        }
        out << answer << endl;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    TaskE solver;
    solver.solve(1, cin, cout);
    return 0;
}
